VEP #233: Enforce Release Branch End-of-Life Policy (#234)
kubevirt/enhancements: veps/sig-release/233-release-branch-eol-policy/vep.md

# VEP #233: Enforce Release Branch End-of-Life Policy

> ## Release Signoff Checklist

Reviewer: It looks like the VEP template has been updated to include a [new metadata section](https://github.com/kubevirt/enhancements/blob/main/veps/NNNN-vep-template/vep.md) and this has changed a little bit as well. I imagine it's worth rebasing to match.

Author: Done, added the new VEP Status Metadata section adapted for this process VEP with announcement (v1.9) and enforcement (v1.10) target versions.

4. **Backport rejection**: Cherry-pick PRs targeting EOL branches are
   auto-closed by a bot with a standard message directing users to upgrade.
5. **Announcement**: EOL is announced on the `kubevirt-dev` mailing list as
   part of the new release announcement.

### CVE Handling for EOL Releases

CVE reports against EOL releases will be handled as follows:

- If the CVE affects a supported release, it will be fixed in supported
  releases only.
- The CVE advisory will note which releases contain the fix and recommend
  upgrading from EOL releases.
- No backports to EOL branches will be made, regardless of severity.

This is consistent with how Kubernetes and most major open-source projects
handle CVEs in EOL releases.

### Downstream Extended Support

Downstream distributors who need longer support windows are free to maintain
their own forks and backport patches independently. The upstream project's EOL
does not prevent downstream organizations from continuing to ship patches for
older releases in their own products.

### Transition Plan

To avoid disruption, the following transition plan is proposed:

1. **Announce the policy** on `kubevirt-dev@googlegroups.com` and in the

>    release notes for the next GA release after this VEP is accepted.

Reviewer: ```suggestion
1. **Announce the policy** on `kubevirt-dev@googlegroups.com`, slack channels,
social media, website, and in the release notes for the next GA release after this
VEP is accepted.
```
I think that covers all our available channels. This is an important change so it would be good to have it as broadcast as far and loud as we can.

Author: Done, added all available channels.

2. **Grace period**: Provide one full release cycle of notice. For example, if

>    announced with v1.8 GA, the policy takes effect at v1.9 GA.

Reviewer: ```suggestion
   announced with v1.9 GA, the policy takes effect at v1.10 GA.
```
If this is accepted for v1.9 then what I understand is Phase 1 would begin at 1.9 GA in July, with Phase 2 taking effect at 1.10 GA.

Author: Done, updated the timeline to announce with v1.9 GA and enforce at v1.10 GA.

3. **Initial EOL batch**: At the policy effective date, all releases older than
   N-2 (where N is the current release) are marked EOL simultaneously. As of

>    v1.9 GA, this would mean v1.6 and earlier are EOL.

Reviewer: If my comment above is correct, then I think this would be clearer written as
```suggestion
   v1.10 GA, this would mean v1.7 and earlier are EOL.
```
Otherwise: disregard

Author: Done, updated to v1.10 GA with v1.7 and earlier marked EOL.

# VEP #233: Enforce Release Branch End-of-Life Policy

## VEP Status Metadata

### Target releases

<!--
This VEP is a process/policy change and does not follow the standard
Alpha/Beta/GA feature graduation model. The target version below reflects
when the policy is announced and takes effect.
-->

- This VEP targets announcement for version: v1.9
- This VEP targets enforcement for version: v1.10

### Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue [#233](https://github.com/kubevirt/enhancements/issues/233) created, which links to VEP dir in [kubevirt/enhancements](https://github.com/kubevirt/enhancements) (not the initial VEP PR)
- [ ] (R) Target version is explicitly mentioned and approved

> - [ ] (R) Graduation criteria filled

Reviewer 2: Both seems to be checked ;)

  branches

## Non Goals

- Changing the KubeVirt release cadence (3 releases per year, trailing
  Kubernetes by ~2 months)
- Modifying the backport policy for branches that are within the support window
- Changing the Kubernetes version compatibility matrix
- Deleting old release branches or git tags (historical artifacts are preserved)
- Defining a long-term support (LTS) release model (this could be a future
  VEP if there is demand)

## Definition of Users

- **KubeVirt release maintainers**: Contributors who cut releases, manage CI
  lanes, and process backport PRs
- **KubeVirt contributors**: Developers who must create and shepherd cherry-pick
  PRs for bug fixes and CVEs
- **Downstream distributors**: Organizations (e.g., Red Hat, SUSE) that
  consume KubeVirt releases and may maintain their own extended support windows
  independently
- **End users**: Operators running KubeVirt clusters who need to understand
  which versions receive security and bug fixes

## User Stories

- As a **release maintainer**, I want a bounded number of release branches to
  maintain so that CVE response does not require cherry-picks to 10+ branches.

> - As a **contributor**, I want a clear policy on which branches accept

Reviewer 2: side note: Would be cool to have a command to automatically backport to supported version :)

   years (release-0.4 through release-0.57), immediately:
   - Remove any remaining CI configuration
   - Apply branch protection rules to prevent merges
   - No need to delete branches; they serve as historical record

### Documentation Updates

The following documentation changes are required:

- **`docs/release.md`**: Replace the soft EOL language with the hard policy
  defined in this VEP. Remove the sentence *"The EOL of a KubeVirt release is
  currently not enforced."*
- **`docs/release-branch-backporting.md`**: Add a section clarifying that
  backports are only accepted for releases within the support window.
- **sig-release support matrix**: Ensure EOL releases are clearly marked and
  the support window policy is documented.

## API Examples

N/A — this VEP is a process and policy change, not an API change.

## Alternatives

### Alternative 1: Support 4 most recent releases

Supporting 4 releases instead of 3 would provide a longer ~16-20 month window.
This was considered but rejected because:

- It still represents a significant improvement over the current unbounded
  state
- It would maintain 4 active branches instead of 3, increasing maintenance
  burden by 33%

> - The 3-release model directly mirrors Kubernetes and is well-understood by the

Reviewer 2: Ultimately this the highlight, its is hard to support KubeVirt once we don't have supported Kubernetes for it...
